NFs/bsf/pkg/app/app.go

```diff
@@ -0,0 +1,184 @@
+	wg            sync.WaitGroup
+}
+
+func NewApp(ctx context.Context, cfg *factory.Config, tlsKeyLogPath string) *App {
+	a := &App{
+		cfg:        cfg,
+		ctx:        ctx,
+		tlsKeyPath: tlsKeyLogPath,
+		bsfCtx:     bsfContext.BsfSelf,
+	}
+
+	// Initialize metrics if enabled
+	if a.cfg.AreMetricsEnabled() {
+		sbiMetrics.EnableSbiMetrics()
+
+		features := map[utils.MetricTypeEnabled]bool{utils.SBI: true}
+		customMetrics := make(map[utils.MetricTypeEnabled][]prometheus.Collector)
+
+		var err error
+		if a.metricsServer, err = metrics.NewServer(
+			getInitMetrics(cfg, features, customMetrics), tlsKeyLogPath, logger.MainLog); err != nil {
+			logger.MainLog.Warnf("Failed to create metrics server: %+v", err)
+		}
+	}
+
+	return a
+}
+
+func getInitMetrics(
+	cfg *factory.Config,
+	features map[utils.MetricTypeEnabled]bool,
+	customMetrics map[utils.MetricTypeEnabled][]prometheus.Collector,
+) metrics.InitMetrics {
+	metricsInfo := metrics.Metrics{
+		BindingIPv4: cfg.GetMetricsBindingAddr(),
+		Scheme:      cfg.GetMetricsScheme(),
+		Namespace:   cfg.GetMetricsNamespace(),
+		Port:        cfg.GetMetricsPort(),
+		Tls: metrics.Tls{
+			Key: cfg.GetMetricsCertKeyPath(),
+			Pem: cfg.GetMetricsCertPemPath(),
+		},
+	}
+
+	// Enable business metrics if configured
+	if cfg.AreMetricsEnabled() {
+		businessMetrics.EnableBindingMetrics()
+		businessMetrics.EnableDiscoveryMetrics()
+
+		// Add BSF business metrics
+		if customMetrics == nil {
+			customMetrics = make(map[utils.MetricTypeEnabled][]prometheus.Collector)
+		}
+
+		// Add binding metrics
+		customMetrics[utils.SBI] = append(
+			customMetrics[utils.SBI],
+			businessMetrics.GetBindingHandlerMetrics(cfg.GetMetricsNamespace())...)
+
+		// Add discovery metrics
+		customMetrics[utils.SBI] = append(
+			customMetrics[utils.SBI],
+			businessMetrics.GetDiscoveryHandlerMetrics(cfg.GetMetricsNamespace())...)
+	}
+
+	return metrics.NewInitMetrics(metricsInfo, "bsf", features, customMetrics)
+}
+
+func (a *App) Start() {
+	defer func() {
+		if p := recover(); p != nil {
+			// Print stack for panic to log. Fatalf() will let program exit.
+			logger.MainLog.Fatalf("panic: %v\n%s", p, string(debug.Stack()))
+		}
+	}()
+
+	// Initialize MongoDB connection
+	if err := a.bsfCtx.ConnectMongoDB(); err != nil {
+		logger.MainLog.Warnf("MongoDB connection failed: %+v", err)
+	} else {
+		// Load existing bindings from MongoDB
+		if err := a.bsfCtx.LoadPcfBindingsFromMongoDB(); err != nil {
+			logger.MainLog.Warnf("Failed to load PCF bindings from MongoDB: %+v", err)
+		}
+	}
+
+	// Start cleanup routine for expired and inactive bindings
+	a.bsfCtx.StartCleanupRoutine()
+
+	// Start metrics server if enabled
+	if a.cfg.AreMetricsEnabled() && a.metricsServer != nil {
+		go func() {
+			a.metricsServer.Run(&a.wg)
+		}()
+		logger.MainLog.Infof("BSF metrics server enabled on %s://%s",
+			a.cfg.GetMetricsScheme(), a.cfg.GetMetricsBindingAddr())
+	}
+
+	// Register with NRF
+	go func() {
```

Comment:
It would be better if you can move the NF Reg part to sbi processor, and use singleton to refactor the Processor and Consumer instances like: https://github.dev/free5gc/pcf/blob/main/cmd/main.go

However, keeping the current design is also acceptable, but it needs to remove the use of the goroutine and change the error handling behavior from printing the error log to returning error directly.